### PR TITLE
WIP: Feature: Custom http server

### PR DESCRIPTION
**Pull request by delvedor**
~~This is a working progress, currently the test are not passing because the `uws` API is slightly different from the core http module.~~
We should find a way to create a compatibility layer (eg: enrich the server instance that we are returning) or try do contact the other modules maintainers.

Feedbacks?

Related: https://github.com/fastify/fastify/issues/365

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by delvedor** [comment]
I've removed `uws` from the test suite since the API is too much different, I don't have time to figure out why is not working.

In any case this should be enough to support a custom http server.
@fastify/fastify some feedbacks?

File: docs/Factory.md [hunk 1]
```diff
@@ -104,3 +104,28 @@ are not present on the object, they will be added accordingly:
...
+```js
+const httpServer = handler => {
+  const server = http.createServer((req, res) => {
+    handler(req, res)
+  })
+
+  return server
+}
+
+const fastify = Fastify({ httpServer })
```
Comment (mcollina) [inline docs/Factory.md]: I would call the option `serverFactory` instead.

File: fastify.js [hunk 2]
```diff
@@ -97,7 +97,9 @@ function build (options) {
 
   var server
   const httpHandler = router.lookup.bind(router)
-  if (options.https) {
+  if (options.serverFactory) {
+    server = options.serverFactory(httpHandler)
+  } else if (options.https) {
```
Comment (mcollina) [inline fastify.js]: A better implementation of this code is to have some server factories around and select one based on the options.

File: docs/Factory.md [hunk 3]
```diff
@@ -104,3 +104,28 @@ are not present on the object, they will be added accordingly:
...
       Any user supplied serializer will override the default serializer of the
       corresponding property.
+
+<a name="custom-http-server"></a>
+### `serverFactory`
+You can pass a custom http server to Fastify by using the `serverFactory` option.<br/>
+`serverFactory` is a function that takes an `handler` parameter, which takes the `request` and `response` objects as parameters.
+
+```js
+const serverFactory = handler => {
```
Comment (mcollina) [inline docs/Factory.md]: I think this should accept the options.

Comment (lucamaraschi) [reply docs/Factory.md]: I agree!

File: fastify.js (hunk 2 above)
Comment (lucamaraschi) [reply fastify.js]: You mean as a strategy pattern implementation?

Comment (mcollina) [reply fastify.js]: ssshhh patterns are not cool anymore ;).

**Pull request merged**